stumpy/core.py

```diff
@@ -1971,3 +1971,81 @@ def _idx_to_mp(I, T, m, normalize=True):
     P[I < 0] = np.inf
 
     return P
+
+
+@njit(
+    # "(i8, i8, i8, i8[:], i8)"
+)
+def _get_tiles(m, n_A, n_B, diags, tile_size):
+    """
+    Split distance matrix into tiles of given size
+
+    Parameters
+    ----------
+    m : int
+        Window size
+
+    n_A : ndarray
+        The length of time series `T_A`
+
+    n_B : ndarray
+        The length of time series `T_B`
+
+    diags : ndarray
+        The diagonal indices of interest
+
+    tile_size : int
+        Maximum length of each tile
+
+    Returns
+    -------
+    tiles : ndarray
+        An array of 7-element-arrays representing each tile:
+        The first two values represent offset positions of the tile
+        The second two values represent dimensions of the tile
+        The third two values represent diagonal indices to traverse in the tile
+        The seventh value represents number of distances in the tile
+    """
+    # dimensions of distance matrix
+    dim = (n_A - m + 1, n_B - m + 1)
+    # number of tiles
+    tiles_count = np.uint64(np.ceil(dim[0] / tile_size) * np.ceil(dim[1] / tile_size))
+    tiles = np.zeros((tiles_count, 7), dtype=np.int64)
+
+    tile_idx = 0
+    # iterate over each tile
+    for y_offset in range(0, dim[0], tile_size):
+        # height of current tile
+        tile_height = min(tile_size, dim[0] - y_offset)
+        for x_offset in range(0, dim[1], tile_size):
+            # width of current tile
+            tile_width = min(tile_size, dim[1] - x_offset)
+
+            # lower and upper diagonal indices to traverse within tile
+            tile_lower_diag = max(1 - tile_height, diags[0] + y_offset - x_offset)
+            tile_upper_diag = min(tile_width, diags[-1] + 1 + y_offset - x_offset)
+
+            tile_ndist = np.sum(
+                _count_diagonal_ndist(
+                    np.arange(tile_lower_diag, tile_upper_diag),
+                    1,
+                    tile_height,
+                    tile_width,
+                )
+            )
+
+            # offset position of tile
+            tiles[tile_idx][0] = y_offset
+            tiles[tile_idx][1] = x_offset
+            # dimensions of tile
+            tiles[tile_idx][2] = tile_height
+            tiles[tile_idx][3] = tile_width
+            # diagonals of tile to traverse
+            tiles[tile_idx][4] = tile_lower_diag
+            tiles[tile_idx][5] = tile_upper_diag
+            # weight of tile
+            tiles[tile_idx][6] = tile_ndist
+
+            tile_idx += 1
+
+    return tiles
```

Comment:
On further thought, I was able to refactor this function into pure numpy:

```
@njit()
def _get_tiles(m, n_A, n_B, diags, tile_size):
    height = n_A - m + 1
    width = n_B - m + 1
    height_count = int(np.ceil(height / tile_size))
    width_count = int(np.ceil(width / tile_size))

    height_offsets = np.repeat(np.arange(0, height, tile_size), width_count)
    width_offsets = np.array(list(np.arange(0, width, tile_size)) * height_count)
    tile_heights = np.minimum(height - height_offsets, tile_size)
    tile_widths = np.minimum(width - width_offsets, tile_size)
    tile_lower_diags = 1 - tile_heights
    tile_upper_diags = tile_widths
    tile_ndists = np.multiply(tile_heights, tile_widths)
    
    return np.column_stack(
        (
            height_offsets,
            width_offsets,
            tile_heights,
            tile_widths,
            tile_lower_diags,
            tile_upper_diags,
            tile_ndists
        )
    )
```

If you don't mind, I think you can take your use your current `for-loop` implementation as the `naive` implementation. Please confirm and test that this function does exactly the same thing as your original function and that I have no missed anything.

Additionally, after rewriting it, I noticed that `tile_upper_diags = tile_widths`. This seems redundant but maybe it's okay?

Reply:
In fact, rather than returning a stacked array, maybe it is okay to return seven 1D arrays because they each have very specific meaning.

`tiles[;, 0]` is meaningless and hard to follow

Reply:
Thanks for rewriting this, looks good to me. The only thing I'd say is once again we're just doing `tile_ndists = np.multiply(tile_heights, tile_widths)` instead of actually computing `tile_ndists` (which is probably fine?).

> If you don't mind, I think you can take your use your current for-loop implementation as the naive implementation.

Yepp that's a good idea, I can do that.

> In fact, rather than returning a stacked array, maybe it is okay to return seven 1D arrays because they each have very specific meaning.

Yepp we can do that. I was returning an array of seven-element arrays earlier because it felt natural to have our first axis as each individual "tiles".

Reply:
> The only thing I'd say is once again we're just doing tile_ndists = np.multiply(tile_heights, tile_widths) instead of actually computing tile_ndists (which is probably fine?).

I don't think I agree with this assessment but maybe I'm missing something. If I set some arbitrary values and do:

```
import numpy as np
height = 10
width_count = 3
tile_size = 4
height_offsets = np.repeat(np.arange(0, height, tile_size), width_count)
# array([0, 0, 0, 4, 4, 4, 8, 8, 8])
tile_heights = np.minimum(height - height_offsets, tile_size)
# array([4, 4, 4, 4, 4, 4, 2, 2, 2])
```
Notice that the height of the last 3 tiles are all `2, 2, 2`, which is what I would expect (i.e., they are smaller because they are at the end). Similarly, the `tile_widths` should also be smaller where necessary. So, I believe that these `tile_heights` and `tile_widths` are reflecting the true square/rectangular dimensions of that tile, which then means that `tile_ndists` is also reflecting the true number of distances being computed in each square/rectangular tile. Can you please confirm?

Reply:
@alvii147 This should work like the original:

```
@njit()
def _get_tiles(m, n_A, n_B, diags, tile_size):
    height = n_A - m + 1
    width = n_B - m + 1
    height_count = int(np.ceil(height / tile_size))
    width_count = int(np.ceil(width / tile_size))

    height_offsets = np.repeat(np.arange(0, height, tile_size), width_count)
    width_offsets = np.array(list(np.arange(0, width, tile_size)) * height_count)
    tile_heights = np.minimum(height - height_offsets, tile_size)
    tile_widths = np.minimum(width - width_offsets, tile_size)
    tile_lower_diags = np.maximum(1 - tile_heights, diags[0] + height_offsets - width_offsets)
    tile_upper_diags = np.minimum(tile_widths, diags[-1] + 1 + height_offsets - width_offsets)

    tile_ndists = np.empty(len(tile_widths), dtype=np.int64)
    tmp_ones = np.ones((tile_size, tile_size))
    for i, (tile_height, tile_width, tile_lower_diag) in enumerate(zip(tile_heights, tile_widths, tile_lower_diags)):
        tile_ndists[i] = np.sum(np.triu(tmp_ones[:tile_height, :tile_width], tile_lower_diag))
        
    return np.column_stack(
        (
            height_offsets,
            width_offsets,
            tile_heights,
            tile_widths,
            tile_lower_diags,
            tile_upper_diags,
            tile_ndists
        )
    )
```

Reply:
@alvii147 If you pull the latest commit then you can access the new function `core._total_diagonal_ndists` via:

```
@njit(fastmath=True, parallel=True)
def _get_tiles(m, n_A, n_B, diags, tile_size):
    height = n_A - m + 1
    width = n_B - m + 1
    height_count = int(np.ceil(height / tile_size))
    width_count = int(np.ceil(width / tile_size))

    height_offsets = np.repeat(np.arange(0, height, tile_size), width_count)
    width_offsets = np.array(list(np.arange(0, width, tile_size)) * height_count)
    tile_heights = np.minimum(height - height_offsets, tile_size)
    tile_widths = np.minimum(width - width_offsets, tile_size)
    tile_lower_diags = np.maximum(1 - tile_heights, diags[0] + height_offsets - width_offsets)
    tile_upper_diags = np.minimum(tile_widths, diags[-1] + 1 + height_offsets - width_offsets)

    tile_ndists = np.empty(len(tile_widths), dtype=np.int64)
    for i in prange(len(tile_ndists)):
        tile_ndists[i] = core._total_diagonal_ndists(tile_lower_diags[i], tile_upper_diags[i], tile_heights[i], tile_widths[i])
        
    return np.column_stack(
        (
            height_offsets,
            width_offsets,
            tile_heights,
            tile_widths,
            tile_lower_diags,
            tile_upper_diags,
            tile_ndists
        )
    )
```